### PR TITLE
Icalendar highlevel interface

### PR DESCRIPTION
**Pull request by matthiasbeyer**
Here comes the icalendar high level interface.

I squashed my own development branch down to 5 commits (it was ~15 before) for better review (and because you squash-merge anyways...)

Tell me what you think!

### CODE REVIEW

**Comment by matthiasbeyer** [comment]
Something weird happened during rebase. The fixup commit should work now.

File: .travis.yml [hunk 1]
```diff
@@ -13,6 +13,11 @@ rust:
   - stable
   - beta
 
+script:
+  - |
+      cargo build --features timeconversions &&
+      cargo test  --features timeconversions
```
Comment (untitaker) [inline .travis.yml]: Weird. Can't we just use one YAML array item for each command?

File: src/error.rs [hunk 2]
```diff
@@ -16,6 +20,10 @@ error_chain! {
             display("Passed content string is not a VCard")
         }
 
+        NotAnIcalendar(content: String) {
```
Comment (untitaker) [inline src/error.rs]: Please capitalize consistently with `VCard`, I.e. `ICalendar`.

File: src/error.rs [hunk 3]
```diff
@@ -16,6 +20,10 @@ error_chain! {
...
+            description("Input is not a valid ICalendar")
+            display("Not an Icalendar: '{}'", content)
```
Comment (untitaker) [inline src/error.rs]: In this error msg as well

File: src/icalendar.rs [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+
+impl Icalendar {
+
+    /// Parse a string to a Icalendar object
+    ///
+    /// Returns an error if the parsed text is not a Icalendar (that means that an error is
+    /// returned also if this is a valid vcard!)
+    ///
+    pub fn build(s: &str) -> Result<Icalendar> {
+        parse_component(s)
```
Comment (untitaker) [inline src/icalendar.rs]: You can use `?` here instead of this "callback chain".

File: src/icalendar.rs [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+            Err(_) => NaiveDate::parse_from_str(&self.0, DATE_FMT)
+                .map(Time::Date)
+                .map_err(From::from),
+        }
+    }
+
+}
+
+#[cfg(feature = "timeconversions")]
+impl AsDateTime for Dtstamp {
```
Comment (untitaker) [inline src/icalendar.rs]: Can't we use `TryInto` or similar here?

File: .travis.yml (hunk 1 above)
Comment (matthiasbeyer) [reply .travis.yml]: You mean as in

```yaml
script:
    - cargo build --features timeconversions
    - cargo test  --features timeconversions
```

? Sure...

**Comment by untitaker** [comment]
Yes, iirc the errors etc look better then (because it shows the output of each command separately)

On 14 November 2017 14:56:38 GMT+01:00, Matthias Beyer <notifications@github.com> wrote:
>matthiasbeyer commented on this pull request.
>
>
>
>> @@ -13,6 +13,11 @@ rust:
>   - stable
>   - beta
> 
>+script:
>+  - |
>+      cargo build --features timeconversions &&
>+      cargo test  --features timeconversions
>
>You mean as in
>
>```yaml
>script:
>    - cargo build --features timeconversions
>    - cargo test  --features timeconversions
>```
>
>? Sure...
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/untitaker/rust-vobject/pull/19#discussion_r150838350

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

File: src/icalendar.rs (hunk 5 above)
Comment (matthiasbeyer) [reply src/icalendar.rs]: Because `TryInto` is not stable yet.

**Comment by untitaker** [comment]
Any updates?

**Comment by matthiasbeyer** [comment]
Sorry, had too much to do this week. Here are the requested fixes.

**Comment by untitaker** [comment]
Awesome thank you so much! Should I release or are there more PRs you want to make?

**Comment by matthiasbeyer** [comment]
I have one small PR I would like to include either way, after that we can discuss whether release or not.

**Pull request merged**